### DISCUSSION TITLE
Slider change event doesn’t fire when slider value set

### DISCUSSION THREAD

**Discussion by dougcooper** [slint-ui/slint#2202] [Q&A] [answered]
I'm trying to change the value of a slider programmatically but the slider change event doesn't fire when I do, nor does the ui update. The underlying value does change. I have a property bound to the slider change event that I'm using in the rust side to get/set the slider value.

**Comment by hunger** [comment]
You set the value from Rust? Did you trigger the event loop after setting the value?

**Reply by dougcooper** [reply]
i did trigger the event loop. everything else works as expected. for instance if i manually drag the slider the event fires but when setting the slider value from rust it does not.

**Reply by hunger** [reply]
I mean did you trigger the event loop after setting the value from Rust?

Try doing

```rust
let _ = slint_interpreter::invoke_from_event_loop(|| {}); // wake event loop
```

after setting the value(s) from within rust to let Slint know that its time to go and update things.

**Reply by dougcooper** [reply]
I tried the code as your wrote it but it did not affect the outcome. Perhaps I need to set the value instead the closure?

**Reply by dougcooper** [reply]
This is how I invoked setting the slider value.

![image](https://user-images.githubusercontent.com/17769091/217547233-f38a1b72-741f-470e-a974-c4fca24a4473.png)

**Comment by ogoffart** [comment, marked as answer]
Indeed, the slider `changed` callback will only be called if the value is set by the user. Programmatically changing the value will not result in this being called.

However, the UI should refresh. If it doesn't that's a bug.
Do you have an example of code that show the problem?

**Reply by dougcooper** [reply]
The ui does not update. It's unfortunate setting the value won't trigger a callback which is a showstopper for me. What im ultimately trying to do is measure the time between when the value is set and when the change call back fires to get an estimate of render time. We are running a ui on an embedded device in csharp using Skia and I wanted a rust example for performance comparison. This is probably a naive strategy given I have no background in graphics. 

I can try to gin up a demo.

**Reply by dougcooper** [reply]
example: https://github.com/dougcooper/slint-slider-test.git

**Reply by dougcooper** [reply]
@ogoffart i had a chance to play with a few other ui frameworks this week and i now have a better understanding of why you wouldnt fire a callback when the slider value changes programmatically. I think ultimately what i wanted to see is draw/render time when the slider changes positions as this appears to be an expensive operation elsewhere. With that said the UI still does not update when the slider value is set as least in my program. i think im doing it right.

**Reply by ogoffart** [reply]
I looked at your code code, and saw this: 

https://github.com/dougcooper/slint-slider-test/blob/88a7fc28154b5855dd24830f9d9925e37b0a1e24/src/main.rs#LL9C33-L9C45

> `in-out property <float> slider-value: slider.value ;`

But the `: slider.value`  is a binding that is only working one way.
This is the same as saying  ` let mut foobar = foo.bar.clone(); `  then changing foobar doesn't change the original value.

The fix is to do

`in-out property <float> slider-value <=> slider.value ;`

Which then does a two way binding, so changing slider-value also affect slider.value.  and then the example works.

**Reply by dougcooper** [reply]
i really appreciate you looking at that. works great. is there a use case where you would want to say a property is an `in` type but not have it bind to the value?

**Reply by ogoffart** [reply]
is there a use case where you would want to say a property is an in type but not have it bind to the value?

To give a default value.
